### PR TITLE
feat: implement network selection

### LINKED ISSUE #25: On different networks (`mainnet` vs `testnet`) and environments (`production` vs `staging`)
## Requirements

1. Not every swap provider supports every network
2. It needs to be clear to users, which network they are currently using
3. We should be able to test new code on mainnet before making it generally available

## Assumptions

1. We don't automatically deploy to production as things are merged into main but instead, this is a dedicated step.

## Proposal

1. The current network is a runtime configuration

- I think it is preferable to have a "universal" build of the app that can run in any environment. If we opt to some form of have "staging" environment, a universal build would allow us to promote the exact same artifacts from the staging environment to prod, without requiring a rebuild.
- We will need to add some infrastructure for swap providers to check the current network and signal whether or not they are compatible with that. A react `Context` sounds like the right thing here.

2. The user can select the current network

3. Have `staging.opendex.app` as a CD target for everything that gets merged into main

4. Default the network based on the URL

This would allow us to by default have `mainnet` on `opendex.app` and `testnet` on `staging.opendex.app`.

## Possible usecases

If we implement the above, the following things become possible:

- A swap provider **without a public testnet** (looking at you, Liquid) can test new code (read: merged to main) before making it generally available (read: released). One can test this by simply selecting "mainnet" on `staging.opendex.app`.
- We can leverage netlify's branch-deploys which can deploy the code of a PR before it is merged. This allows us to test a feature before even merging it into main.
- Users can try out / test new features (on `staging.opendex.app`) before they are made generally available. Again, "testing" here is not restricted to using testnet. One can easily test on mainnet.

### PR DESCRIPTION
**Pull request by michael1011**
Closes #25 

This is an early draft for implementing the network selection outlined in #25. I would like some feedback on the approach before I continue with this PR.

TODOs:
- [x] implement selection to switch networks
- [x] default values based on URL

### CODE REVIEW

File: src/context/NetworkContext.tsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      }}
+    >
+      <>{children}</>
+    </NetworkContext.Provider>
+  );
+};
+
+export {
+  Network,
+  NetworkContext,
```
Comment (thomaseizinger) [inline src/context/NetworkContext.tsx:94]: I'd rather export a hook to get the current network. Something like `useNetwork`.

File: src/context/NetworkContext.tsx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    return input.charAt(0).toUpperCase() + input.slice(1);
+  };
+  
+  const boltz: BoltzConfiguration = {
+    apiEndpoint: process.env[`REACT_APP_BOLTZ_${network}_API`]!,
+    infuraId: process.env[`REACT_APP_BOLTZ_${network}_INFURA_ID`]!,
+
+    bitcoinConstants: Networks[`bitcoin${capitalizeFirstLetter(network.toLowerCase())}`],
+    litecoinConstants: Networks[`litecoin${capitalizeFirstLetter(network.toLowerCase())}`],
+  };
```
Comment (thomaseizinger) [inline src/context/NetworkContext.tsx:60]: I think a more scalable way would be to remove this functionality from here and write a custom hook that composes `useNetwork` (to be defined based on the below context) along the lines of `useBoltzConfiguration`. This allows us to have this functionality per swap provider and we don't have to all edit the same component.

File: .env [hunk 3]
```diff
@@ -1,30 +1,48 @@
...
-# Onion URL
-REACT_APP_BOLTZ_ONION=http://random.onion
+# Block explorers
 
-# API endpoint
-REACT_APP_BOLTZ_API=https://testnet.boltz.exchange/api
-REACT_APP_BOLTZ_API_ONION=http://tboltzzrsoc3npe6sydcrh37mtnfhnbrilqi45nao6cgc6dr7n2eo3id.onion
+# Mainnet
+REACT_APP_MAINNET_EXPLORER_BTC_ADDRESS=https://blockstream.info/address/
+REACT_APP_MAINNET_EXPLORER_BTC_TRANSACTION=https://blockstream.info/tx/
```
Comment (thomaseizinger) [inline .env:9]: Outside of this PRs scope: Have you considered only declaring the env variable for the hostname and appending `/address` and `/tx` inside the app? Would cut down the number of variables by quite a bit.

**Review by thomaseizinger — COMMENTED** [review]
Thanks for pushing this forward!

I think a more hook-oriented approach would scale better as we need to add more configuration for different providers.

File: .env (hunk 3 above)
Comment (michael1011) [reply .env:9]: Yes but what if we would like to use an explorer that does not adhere to the `/address` and `/tx` standard? [The litecoin testnet explorer we are using does not for example](https://github.com/opendexnetwork/opendex.app/pull/26/files#diff-e9cbb0224c4a3d23a6019ba557e0cd568c1ad5e1582ff1e335fb7d99b7a1055dR21).

For the sake of keeping the environment variable parsing dead simple, I left it the way it is right now. I don't see the current way as problematic.

Comment (thomaseizinger) [reply .env:9]: > Yes but what if we would like to use an explorer that does not adhere to the /address and /tx standard? The litecoin testnet explorer we are using does not for example.

A very valid reason to have different env variables!

> For the sake of keeping the environment variable parsing dead simple, I left it the way it is right now. I don't see the current way as problematic.

Not at all problematic, I was primarily curious what the reasoning was :)

**Comment by michael1011** [comment]
@thomaseizinger is this what you had in mind?

File: src/context/NetworkContext.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  for (const currency of Object.keys(CurrencyID)) {
+    const blockExplorer: BlockExplorerConfiguration = {
+      address: process.env[`REACT_APP_${network}_EXPLORER_${currency}_ADDRESS`]!,
+      transaction: process.env[`REACT_APP_${network}_EXPLORER_${currency}_TRANSACTION`]!,
+    };
+
+    if (blockExplorer.address !== undefined && blockExplorer.transaction !== undefined) {
+      explorers.set(currency as CurrencyID, blockExplorer);
+    }
+  }
```
Comment (thomaseizinger) [inline src/context/NetworkContext.tsx:72]: At least COMIT is not going to need this in the near future :)

Doesn't hurt to still have it here but it is an indicator that even just the `network` value itself is useful and the block explorer URLs could be composed on top of this.

**Review by thomaseizinger — COMMENTED** [review]
@michael1011 

> @thomaseizinger is this what you had in mind?

Yes! Thanks for incorporating it!

File: src/context/NetworkContext.tsx (hunk 4 above)
Comment (michael1011) [reply src/context/NetworkContext.tsx:72]: So block explorers into their own hook? Fine by me

**Review by tiero — APPROVED** [review]
utACK 

will add liquid env after this got merged

**Review by thomaseizinger — APPROVED** [review]
ACK

Thanks for driving this forward!

**Comment by michael1011** [comment]
I'll make the CSS a little prettier and afterwards merge this PR

**Pull request merged**